src/solve/fulfill.rs

```diff
@@ -115,152 +133,235 @@ impl<'s> Fulfill<'s> {
+            fulfill.push_goal(&goal.environment, goal.goal);
+
+            // Negate the result
+            if let Ok(solution) =  fulfill.solve(Substitution::empty()) {
+                match solution {
+                    Solution::Unique(_) => Err("refutation failed")?,
+                    Solution::Ambig(_) => return Ok(Outcome::Incomplete),
+                }
+            }
+        }
+
+        Ok(Outcome::Complete)
+    }
+
+    /// Try to fulfill all pending obligations. The returned solution will
+    /// transform `subst` substitution with the outcome of type inference by
+    /// updating the replacements it provides.
+    pub fn solve(mut self, subst: Substitution) -> Result<Solution> {
+        // We first solve all positive (to_prove) goals, since these are the
+        // only goals that can affect inference state or add new obligations.
+        let positive = self.fulfill_positive()?;
+
+        // Next we look at the goals we must refute; doing so cannot create new
+        // obligations of any form
+        let negative = self.fulfill_negative()?;
+
+        if (positive.is_complete() && negative.is_complete()) && !self.ambiguous {
+            // No obligations remain, so we have definitively solved our goals,
+            // and the current inference state is the unique way to solve them.
+
+            let constraints = self.constraints.into_iter().collect();
+            let constrained = self.infer.canonicalize(&ConstrainedSubst { subst, constraints });
+            return Ok(Solution::Unique(constrained.quantified))
         }
 
-        Ok(solution.successful)
+        // Otherwise, we have (positive or negative) obligations remaining, but
+        // haven't proved that it's *impossible* to satisfy out obligations. we
+        // need to determine how to package up what we learned about type
+        // inference as an ambiguous solution.
+
+        if subst.is_trivial(&mut self.infer) {
+            // In this case, we didn't learn *anything* definitively. So now, we
+            // go one last time through the positive obligations, this time
+            // applying even *tentative* inference suggestions, so that we can
+            // yield these upwards as our own suggestions. In particular, we
+            // yield up the first one we can find.
+
+            while let Some(wc) = self.to_prove.pop() {
+                let (free_vars, solution) = self.solve_obligation(&wc).unwrap();
+                if let Some(constrained_subst) = solution.constrained_subst() {
+                    self.apply_solution(free_vars, constrained_subst);
+                    let subst = self.infer.canonicalize(&subst);
+                    return Ok(Solution::Ambig(Guidance::Suggested(subst.quantified)));
+                }
+            }
+
+            Ok(Solution::Ambig(Guidance::Unknown))
+        } else {
+            // While we failed to prove the goal, we still leared that
```

Comment:
Typo: s/leared/learned/

(I made this comment earlier, in my review, but GH marked it as "outdated" for some reason.)